// @loaders.gl/crypto/dist/index.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
// __VERSION__ is injected by babel-plugin-version-inline
// @ts-ignore TS2304: Cannot find name '__VERSION__'.
const VERSION = typeof "4.2.0-beta.2" !== 'undefined' ? "4.2.0-beta.2" : 'latest';
export { Hash } from "./lib/hash.js";
export { CRC32Hash } from "./lib/crc32-hash.js";
export { CRC32CHash } from "./lib/crc32c-hash.js";
export { MD5Hash } from "./lib/md5-hash.js";
export { SHA256Hash } from "./lib/sha256-hash.js";
export { CryptoHash } from "./lib/crypto-hash.js";
export { NodeHash } from "./lib/node-hash.js";
/**
 * Small, fast worker for CRC32, CRC32c and MD5 Hashes
 */
export const CryptoWorker = {
    id: 'crypto',
    name: 'CRC32, CRC32c and MD5 Hashes',
    module: 'crypto',
    version: VERSION,
    options: {
        crypto: {}
    }
};
/**
 * Large worker for full complement of Cryptographic Hashes
 * bundles the full crypto.js library
 */
export const CryptoJSWorker = {
    id: 'cryptojs',
    name: 'Cryptographic Hashes',
    module: 'crypto',
    version: VERSION,
    options: {
        cryptojs: {}
    }
};
// EXPERIMENTAL
export { encodeNumber, encodeHex, encodeBase64 } from "./lib/utils/digest-utils.js";
export { asciiToBase64, base64ToAscii } from "./lib/utils/base64-utils.js";
