// @loaders.gl/math/dist/index.js
export { GL } from "./geometry/constants.js";
// GL support
export { GL_TYPE } from "./geometry/constants.js";
export { default as GLType } from "./geometry/gl/gl-type.js";
// Geometry
export { default as isGeometry } from "./geometry/is-geometry.js";
// Iterators
export { makeAttributeIterator } from "./geometry/iterators/attribute-iterator.js";
export { makePrimitiveIterator } from "./geometry/iterators/primitive-iterator.js";
// Helper methods
export { computeVertexNormals } from "./geometry/attributes/compute-vertex-normals.js";
export { encodeRGB565, decodeRGB565 } from "./geometry/colors/rgb565.js";
// Typed array utils
export { concatTypedArrays } from "./geometry/typed-arrays/typed-array-utils.js";
// Compression
export { octEncodeInRange, octEncode, octEncodeToVector4, octDecodeInRange, octDecode, octDecodeFromVector4, octPackFloat, octEncodeFloat, octDecodeFloat, octPack, octUnpack, compressTextureCoordinates, decompressTextureCoordinates, zigZagDeltaDecode } from "./geometry/compression/attribute-compression.js";
export { emod } from "./geometry/utils/coordinates.js";
